File: 2022/day4/day4.py
import sys
ssum=0
sum=0
for line in open(sys.argv[1]):
	a=line[:line.find(",")]
	b=line[line.find(",")+1:]
	r1=[int(a[:a.find("-")]),int(a[a.find("-")+1:])]
	r2=[int(b[:b.find("-")]),int(b[b.find("-")+1:])]
	if(r1[1]>=r2[0] and r1[0]<=r2[1]):
	    ssum+=1
	    if((r1[0]<=r2[0] and r1[1]>=r2[1]) or (r2[0]<=r1[0] and r2[1]>=r1[1])):
	        sum+=1
print(sum)
print(ssum)

File: 2022/day4/day4.c
#include <stdio.h>
#include <sys/mman.h>
#include <fcntl.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/stat.h>
#define isDigit() (*region>='0')
#define populate(var) \
  var=*(region++)-'0';\
  if(isDigit()) var=var*10+*(region++)-'0';\
  region++


char str[10] = "000000000\n"; 

char* itoa(int n) {
  uint i = 8;
  for(; n > 9; i--) {
    str[i] = (n % 10) + '0';
    n /= 10;
  }
  str[i] = n % 10 + '0';
  return str + i;
}
int main(int argc, char* argv[]) {
  register short f1,f2,s1,s2;
  register int sum=0;
  register int ssum=0;
  {
    auto int fd = open(argv[1], O_RDONLY);
    auto char* region;
    {
      auto struct stat sb;
      fstat(fd,&sb);       
      region = mmap(NULL,
                    sb.st_size,
                    PROT_READ,
                    MAP_PRIVATE,
                    fd,0);
    }
    while(*region != 0) {
      populate(f1);
      populate(f2);
      populate(s1);
      populate(s2);
      if(f2 >= s1 && f1 <= s2) {
        ssum++;
        if((f1 <= s1 && f2 >= s2) || (s1 <= f1 && s2 >= f2))
          sum++;
      }
    }
    close(fd);
  }
  puts("First half:");
  write(1,itoa(sum),10);
  puts("Second half:");
  write(1,itoa(ssum),10);
  return 0;
}
